In [1]:
#Reading the csv files into dataframes
import pandas as pd
import numpy as np
CabData = pd.read_csv("Cab_Data.csv")
City = pd.read_csv("City.csv")
Customers = pd.read_csv("Customer_ID.csv")
Transactions = pd.read_csv("Transaction_ID.csv")

In [2]:
CabData
CabData['Date of Travel'] = pd.to_datetime(CabData['Date of Travel'])
CabData.info()
CabData.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 359392 entries, 0 to 359391
Data columns (total 7 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   Transaction ID  359392 non-null  int64         
 1   Date of Travel  359392 non-null  datetime64[ns]
 2   Company         359392 non-null  object        
 3   City            359392 non-null  object        
 4   KM Travelled    359392 non-null  float64       
 5   Price Charged   359392 non-null  float64       
 6   Cost of Trip    359392 non-null  float64       
dtypes: datetime64[ns](1), float64(3), int64(1), object(2)
memory usage: 19.2+ MB


Transaction ID    0
Date of Travel    0
Company           0
City              0
KM Travelled      0
Price Charged     0
Cost of Trip      0
dtype: int64

In [3]:
#general transaction details of the cab companies
CabData

Transaction ID Date of Travel     Company           City  \
0             10000011     2016-08-01    Pink Cab     ATLANTA GA   
1             10000012     2016-06-01    Pink Cab     ATLANTA GA   
2             10000013     2016-02-01    Pink Cab     ATLANTA GA   
3             10000014     2016-07-01    Pink Cab     ATLANTA GA   
4             10000015     2016-03-01    Pink Cab     ATLANTA GA   
...                ...            ...         ...            ...   
359387        10440101     2018-08-01  Yellow Cab  WASHINGTON DC   
359388        10440104     2018-04-01  Yellow Cab  WASHINGTON DC   
359389        10440105     2018-05-01  Yellow Cab  WASHINGTON DC   
359390        10440106     2018-05-01  Yellow Cab  WASHINGTON DC   
359391        10440107     2018-02-01  Yellow Cab  WASHINGTON DC   

        KM Travelled  Price Charged  Cost of Trip  
0              30.45         370.95      313.6350  
1              28.62         358.52      334.8540  
2               9.04         125.20       97.6320  
3              33.17         377.40      351.6020  
4               8.73         114.62       97.7760  
...              ...            ...           ...  
359387          4.80          69.24       63.3600  
359388          8.40         113.75      106.8480  
359389         27.75         437.07      349.6500  
359390          8.80         146.19      114.0480  
359391         12.76         191.58      177.6192  

[359392 rows x 7 columns]

In [4]:
#Users of the cab services in relation to the Cities
City
#City['Users'] = City['Users'].astype(str).astype(int)
#convert the datatypes from the csv file to enable arithmetic calculations
City.info()
City.isna().sum()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   City        20 non-null     object
 1   Population  20 non-null     int64 
 2   Users       20 non-null     int64 
dtypes: int64(2), object(1)
memory usage: 608.0+ bytes


City          0
Population    0
Users         0
dtype: int64

In [5]:
#Customer details
Customers
Customers.isna().sum()

Customer ID           0
Gender                0
Age                   0
Income (USD/Month)    0
dtype: int64

In [6]:
#Individual transactions
Transactions
Transactions.isna().sum()

Transaction ID    0
Customer ID       0
Payment_Mode      0
dtype: int64

In [7]:
#merging the cab data 
Transactions_Per_city = pd.merge(CabData, Transactions, how='inner')
Transactions_Per_city.dropna(inplace = True)

Transactions_Per_city

Transaction ID Date of Travel     Company           City  \
0             10000011     2016-08-01    Pink Cab     ATLANTA GA   
1             10000012     2016-06-01    Pink Cab     ATLANTA GA   
2             10000013     2016-02-01    Pink Cab     ATLANTA GA   
3             10000014     2016-07-01    Pink Cab     ATLANTA GA   
4             10000015     2016-03-01    Pink Cab     ATLANTA GA   
...                ...            ...         ...            ...   
359387        10440101     2018-08-01  Yellow Cab  WASHINGTON DC   
359388        10440104     2018-04-01  Yellow Cab  WASHINGTON DC   
359389        10440105     2018-05-01  Yellow Cab  WASHINGTON DC   
359390        10440106     2018-05-01  Yellow Cab  WASHINGTON DC   
359391        10440107     2018-02-01  Yellow Cab  WASHINGTON DC   

        KM Travelled  Price Charged  Cost of Trip  Customer ID Payment_Mode  
0              30.45         370.95      313.6350        29290         Card  
1              28.62         358.52      334.8540        27703         Card  
2               9.04         125.20       97.6320        28712         Cash  
3              33.17         377.40      351.6020        28020         Cash  
4               8.73         114.62       97.7760        27182         Card  
...              ...            ...           ...          ...          ...  
359387          4.80          69.24       63.3600        52392         Cash  
359388          8.40         113.75      106.8480        53286         Cash  
359389         27.75         437.07      349.6500        52265         Cash  
359390          8.80         146.19      114.0480        52175         Card  
359391         12.76         191.58      177.6192        52917         Card  

[359392 rows x 9 columns]

In [8]:
Transactions_Per_city.to_csv('Transactions_Per_city.csv', sep = ',', index=True)

In [9]:
#projecting only the transactions and their payment modes in each city
Transactions_Per_city[['City','Payment_Mode']]

City Payment_Mode
0          ATLANTA GA         Card
1          ATLANTA GA         Card
2          ATLANTA GA         Cash
3          ATLANTA GA         Cash
4          ATLANTA GA         Card
...               ...          ...
359387  WASHINGTON DC         Cash
359388  WASHINGTON DC         Cash
359389  WASHINGTON DC         Cash
359390  WASHINGTON DC         Card
359391  WASHINGTON DC         Card

[359392 rows x 2 columns]

In [10]:
#showing card transaction data
card_payments =Transactions_Per_city.loc[Transactions_Per_city['Payment_Mode'] == 'Card', ['City', 'Payment_Mode']]
card_payments



City Payment_Mode
0          ATLANTA GA         Card
1          ATLANTA GA         Card
4          ATLANTA GA         Card
6           AUSTIN TX         Card
7           AUSTIN TX         Card
...               ...          ...
359382  WASHINGTON DC         Card
359385  WASHINGTON DC         Card
359386  WASHINGTON DC         Card
359390  WASHINGTON DC         Card
359391  WASHINGTON DC         Card

[215504 rows x 2 columns]

In [11]:
#showing cash transaction payment
cash_payments =Transactions_Per_city.loc[Transactions_Per_city['Payment_Mode'] == 'Cash', ['City', 'Payment_Mode']]
cash_payments

City Payment_Mode
2          ATLANTA GA         Cash
3          ATLANTA GA         Cash
5          ATLANTA GA         Cash
8           BOSTON MA         Cash
9           BOSTON MA         Cash
...               ...          ...
359383  WASHINGTON DC         Cash
359384  WASHINGTON DC         Cash
359387  WASHINGTON DC         Cash
359388  WASHINGTON DC         Cash
359389  WASHINGTON DC         Cash

[143888 rows x 2 columns]

In [12]:
#difference in cost and price charged for cash transactions
charges=Transactions_Per_city.loc[Transactions_Per_city['Payment_Mode'] == 'Cash', ["Price Charged","Cost of Trip"]]
charges



Price Charged  Cost of Trip
2              125.20       97.6320
3              377.40      351.6020
5               72.43       63.0240
8              191.61      146.8800
9              156.98      113.2560
...               ...           ...
359383         167.09      118.8000
359384         358.77      285.6672
359387          69.24       63.3600
359388         113.75      106.8480
359389         437.07      349.6500

[143888 rows x 2 columns]

In [13]:
#average charges for cash transactions
mean_charges = charges[["Price Charged","Cost of Trip"]].mean()
mean_charges

Price Charged    423.519362
Cost of Trip     286.016438
dtype: float64

In [14]:
#difference in cost and price charged for card transactions
charges2=Transactions_Per_city.loc[Transactions_Per_city['Payment_Mode'] == 'Card', ["Price Charged","Cost of Trip"]]
charges2


Price Charged  Cost of Trip
0              370.95      313.6350
1              358.52      334.8540
4              114.62       97.7760
6              576.15      475.2000
7              466.10      377.8900
...               ...           ...
359382         458.64      472.5120
359385         202.70      159.0732
359386         452.19      351.4104
359390         146.19      114.0480
359391         191.58      177.6192

[215504 rows x 2 columns]

In [15]:
#average charges for cash transactions
mean_charges2 = charges2[["Price Charged","Cost of Trip"]].mean()
mean_charges2

Price Charged    423.392533
Cost of Trip     286.306072
dtype: float64

In [16]:
#process of classifying the transactions by the seasons of the year based on the months of each date of travel
# the seasosn will be classified into
#1 Spring(March to May)
#2 Summer(June to August)
#3 Autumn(September to November)
#4 Winter(December to February)
Seasonality=Transactions_Per_city[['Transaction ID','Date of Travel']]
Seasonality


Transaction ID Date of Travel
0             10000011     2016-08-01
1             10000012     2016-06-01
2             10000013     2016-02-01
3             10000014     2016-07-01
4             10000015     2016-03-01
...                ...            ...
359387        10440101     2018-08-01
359388        10440104     2018-04-01
359389        10440105     2018-05-01
359390        10440106     2018-05-01
359391        10440107     2018-02-01

[359392 rows x 2 columns]

In [17]:
Seasonality.rename(columns = {'Date of Travel':'Date_of_Travel'}, inplace = True)
Seasonality

C:\Users\SIR BOLU\anaconda3\lib\site-packages\pandas\core\frame.py:4441: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


Transaction ID Date_of_Travel
0             10000011     2016-08-01
1             10000012     2016-06-01
2             10000013     2016-02-01
3             10000014     2016-07-01
4             10000015     2016-03-01
...                ...            ...
359387        10440101     2018-08-01
359388        10440104     2018-04-01
359389        10440105     2018-05-01
359390        10440106     2018-05-01
359391        10440107     2018-02-01

[359392 rows x 2 columns]

In [18]:
Spring=Seasonality[(Seasonality.Date_of_Travel.dt.month > 2) & (Seasonality.Date_of_Travel.dt.month < 6)]
Spring

Transaction ID Date_of_Travel
4             10000015     2016-03-01
6             10000017     2016-03-01
23            10000034     2016-03-01
32            10000043     2016-03-01
35            10000046     2016-03-01
...                ...            ...
359384        10440098     2018-05-01
359385        10440099     2018-04-01
359388        10440104     2018-04-01
359389        10440105     2018-05-01
359390        10440106     2018-05-01

[78132 rows x 2 columns]

In [19]:
Summer=Seasonality[(Seasonality.Date_of_Travel.dt.month > 5) & (Seasonality.Date_of_Travel.dt.month < 9)]
Summer

Transaction ID Date_of_Travel
0             10000011     2016-08-01
1             10000012     2016-06-01
3             10000014     2016-07-01
5             10000016     2016-07-01
7             10000018     2016-07-01
...                ...            ...
359374        10440087     2018-06-01
359376        10440089     2018-07-01
359378        10440092     2018-08-01
359386        10440100     2018-07-01
359387        10440101     2018-08-01

[89202 rows x 2 columns]

In [20]:
Autumn=Seasonality[(Seasonality.Date_of_Travel.dt.month > 8) & (Seasonality.Date_of_Travel.dt.month < 12)]
Autumn

Transaction ID Date_of_Travel
10            10000021     2016-11-01
12            10000023     2016-11-01
15            10000026     2016-10-01
24            10000035     2016-09-01
25            10000036     2016-11-01
...                ...            ...
359194        10439882     2018-09-01
359218        10439908     2018-09-01
359233        10439926     2018-09-01
359313        10440021     2018-09-01
359340        10440049     2018-09-01

[108321 rows x 2 columns]

In [21]:
Winter=Seasonality[(Seasonality.Date_of_Travel.dt.month == 12) | (Seasonality.Date_of_Travel.dt.month ==1)| (Seasonality.Date_of_Travel.dt.month ==2)]
Winter

Transaction ID Date_of_Travel
2             10000013     2016-02-01
8             10000019     2016-12-01
18            10000029     2016-02-01
19            10000030     2016-02-01
22            10000033     2016-12-01
...                ...            ...
359344        10440053     2018-02-01
359345        10440054     2018-02-01
359356        10440067     2018-02-01
359379        10440093     2018-12-31
359391        10440107     2018-02-01

[83737 rows x 2 columns]

In [22]:
#Finding the city with the highest per-capita patronage
#We will consider the City dataset
City2=City[['City', 'Population']]
Patronage_per_capita = City["Users"]/City["Population"]
Patronage_per_capita.to_csv('Patronage_per_capita.csv',sep = ',', index=True)
City2

City  Population
0        NEW YORK NY     8405837
1         CHICAGO IL     1955130
2     LOS ANGELES CA     1595037
3           MIAMI FL     1339155
4     SILICON VALLEY     1177609
5      ORANGE COUNTY     1030185
6       SAN DIEGO CA      959307
7         PHOENIX AZ      943999
8          DALLAS TX      942908
9         ATLANTA GA      814885
10         DENVER CO      754233
11         AUSTIN TX      698371
12        SEATTLE WA      671238
13         TUCSON AZ      631442
14  SAN FRANCISCO CA      629591
15     SACRAMENTO CA      545776
16     PITTSBURGH PA      542085
17     WASHINGTON DC      418859
18      NASHVILLE TN      327225
19         BOSTON MA      248968

In [23]:
Patronage = pd.read_csv('Patronage_per_capita.csv')
Patronage=Patronage.iloc[:, [1]]
Patronage.rename(columns = {'0':'Patronage_per_capita'}, inplace = True)
Patronage

Patronage_per_capita
0               0.035945
1               0.084121
2               0.090363
3               0.013199
4               0.023138
5               0.012613
6               0.072964
7               0.006497
8               0.023499
9               0.030312
10              0.016468
11              0.021447
12              0.037338
13              0.009046
14              0.339282
15              0.012906
16              0.006720
17              0.303207
18              0.028329
19              0.321411

In [31]:
#City3 = City2.append(Patronage, ignore_index = False)
City3=pd.concat([City2, Patronage], axis=1)
City3.to_csv('City3.csv',sep = ',', index=True)
City3.sort_values(by=['Patronage_per_capita'], inplace=True, ascending= False)
City3

City  Population  Patronage_per_capita
14  SAN FRANCISCO CA      629591              0.339282
19         BOSTON MA      248968              0.321411
17     WASHINGTON DC      418859              0.303207
2     LOS ANGELES CA     1595037              0.090363
1         CHICAGO IL     1955130              0.084121
6       SAN DIEGO CA      959307              0.072964
12        SEATTLE WA      671238              0.037338
0        NEW YORK NY     8405837              0.035945
9         ATLANTA GA      814885              0.030312
18      NASHVILLE TN      327225              0.028329
8          DALLAS TX      942908              0.023499
4     SILICON VALLEY     1177609              0.023138
11         AUSTIN TX      698371              0.021447
10         DENVER CO      754233              0.016468
3           MIAMI FL     1339155              0.013199
15     SACRAMENTO CA      545776              0.012906
5      ORANGE COUNTY     1030185              0.012613
13         TUCSON AZ      631442              0.009046
16     PITTSBURGH PA      542085              0.006720
7         PHOENIX AZ      943999              0.006497

In [25]:
#Finding the most popular cab company in each city
#1. First create two dataframes of each of the Cab companies
#2. Find the most popular Cab company by raw numbers
#3. Classify them by City

Transactions_Per_city

Transaction ID Date of Travel     Company           City  \
0             10000011     2016-08-01    Pink Cab     ATLANTA GA   
1             10000012     2016-06-01    Pink Cab     ATLANTA GA   
2             10000013     2016-02-01    Pink Cab     ATLANTA GA   
3             10000014     2016-07-01    Pink Cab     ATLANTA GA   
4             10000015     2016-03-01    Pink Cab     ATLANTA GA   
...                ...            ...         ...            ...   
359387        10440101     2018-08-01  Yellow Cab  WASHINGTON DC   
359388        10440104     2018-04-01  Yellow Cab  WASHINGTON DC   
359389        10440105     2018-05-01  Yellow Cab  WASHINGTON DC   
359390        10440106     2018-05-01  Yellow Cab  WASHINGTON DC   
359391        10440107     2018-02-01  Yellow Cab  WASHINGTON DC   

        KM Travelled  Price Charged  Cost of Trip  Customer ID Payment_Mode  
0              30.45         370.95      313.6350        29290         Card  
1              28.62         358.52      334.8540        27703         Card  
2               9.04         125.20       97.6320        28712         Cash  
3              33.17         377.40      351.6020        28020         Cash  
4               8.73         114.62       97.7760        27182         Card  
...              ...            ...           ...          ...          ...  
359387          4.80          69.24       63.3600        52392         Cash  
359388          8.40         113.75      106.8480        53286         Cash  
359389         27.75         437.07      349.6500        52265         Cash  
359390          8.80         146.19      114.0480        52175         Card  
359391         12.76         191.58      177.6192        52917         Card  

[359392 rows x 9 columns]

In [26]:
#profits of each company over the time period
#1. First create two dataframes of each of the Cab companies
#2. Find the the difereence between the cost of travel and the price charged
#3. Classify them by City

Income= Transactions_Per_city[['Price Charged','Cost of Trip']]
Income
#Overall profit for each transaction
#Rename the columns to more suitable naming standards
Income.rename(columns = {'Price Charged':'Price_Charged', 'Cost of Trip':'Cost_of_Trip'}, inplace = True)
Profit = Income["Price_Charged"]-Income["Cost_of_Trip"]
Profit

#Total overall profit for both cabs
Total = Profit.sum()
print (Total)

49327701.491799995


C:\Users\SIR BOLU\anaconda3\lib\site-packages\pandas\core\frame.py:4441: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


In [27]:
#Classifying profits by cab company
Pink_cab_transactions = Transactions_Per_city.loc[Transactions_Per_city['Company'] == 'Pink Cab', ['Price Charged', 'Cost of Trip']]
Pink_cab_transactions
#Rename the columns to more suitable naming standards
Pink_cab_transactions.rename(columns = {'Price Charged':'Price_Charged', 'Cost of Trip':'Cost_of_Trip'}, inplace = True)
#overall profit for the pink cab
Pink_cab_profit = Pink_cab_transactions["Price_Charged"]-Pink_cab_transactions["Cost_of_Trip"]
Pink_cab_profit = Pink_cab_profit.sum()
print(Pink_cab_profit)

5307328.321


In [28]:
#Classifying profits by cab company
Yellow_cab_transactions = Transactions_Per_city.loc[Transactions_Per_city['Company'] >= 'Yellow Cab', ['Price Charged', 'Cost of Trip']]
Yellow_cab_transactions
#Rename the columns to more suitable naming standards
Yellow_cab_transactions.rename(columns = {'Price Charged':'Price_Charged', 'Cost of Trip':'Cost_of_Trip'}, inplace = True)
#overall profit for the pink cab
Yellow_cab_profit = Yellow_cab_transactions["Price_Charged"]-Yellow_cab_transactions["Cost_of_Trip"]
Yellow_cab_profit = Yellow_cab_profit.sum()
print(Yellow_cab_profit)

44020373.17080001


<center><p><h1>The Report of my findings</h1><center>

<body> 
<p>Data Preparation</p>    
    <p>i)   I read the datasets into pandas dataframes</p>    
    <p>ii)  I then displayed the dataframes </p> 
    <p>iii) I checked to see if there were any null values to be taken care of</p> 
    
HYPOTHESES:
1) <b>Is there a significant difference in the choice of payment type between Card or Cash?</b><br>
I merged the Cab Data with the transaction data and then found out the following;<br>
a) Cards are slightly more popular as payment modes so begs the question, why are they more popular?<br>
  i)is it because it is cheaper to use a card payment?<br>
  ii)is it because it is more convenient?<br>
   
To find out, I had to compare the price charged and the cost of trip for both payment 
modes to see if there is a significant difference<br>	
Step 1: Get the average price charged as well as the total cost of trip for each payment mode<br>
Step 2: Visualize both and compare<br><br>
	
I have found that there is negligible difference between them so I believe that the card 
payments are probably just out of convenience
	
2) <b>Is there a difference between the frequency of cab rides in each Season of the year?</b><br>
I found the following;<br>
    Total transactions in Spring: 78,132 rides<br>
                       in Summer: 89,202 rides<br>
                       in Autumn: 108,321 rides<br>
                       in Winter: 83,737 rides<br>
    This suggested that the cab companies were more popular in the Autumn season over the two years of study<br>

3) <b>What is the company with more profits?</b><br>
To answer the question, I go through a list of steps including:<br>
    i)Take the earlier merged dataset<br>
    ii)Project the columns of the cost of travel and the price paid for that ride<br>
    iii)Classify the data by the cab companies and find the overall difference of those figures<br>
     After going through these steps,I discovered the following:<br>
    Over the period of study the profits for each cab company include:<br>
    Pink Cab:5,307,328 dollars<br>
    Yellow Cab:44,020,373 dollars<br>
    This suggested that the Yellow cab company was more profitable over the two years of study<br>

4) <b>In which cities are the cabs most popular?</b><br>
    I used the City dataframe to answer this question;
    I found the ratio of the Number of Customers to the overall population for  each city and i found that the top 3 cities where the cab companies are being used are: San Fransisco, Boston and Washington DC
</body>
	